gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py

```diff
@@ -690,6 +697,215 @@ def join_release_ht_with_subsets(
+        )
+        hts.append(subset_ht)
+
+    return concatenate_annotations(
+        hts,
+        row_field_names=["group_membership"],
+        global_field_names=list(FREQ_GLOBALS) + ["raw_group"],
+    )
+
+
+def compute_an_by_group_membership(
+    vds: hl.vds.VariantDataset,
+    group_membership_ht: hl.Table,
+    variant_filter_ht: hl.Table,
+) -> hl.Table:
+    """
+    Compute the allele number for new variants in the v4.0 release by call stats group membership.
+
+    :param vds: VariantDataset with all v3.1 samples (including non-release) and
+        multi-allelics split.
+    :param group_membership_ht: Table with the group membership for each sample. This
+        is generated by `get_group_membership_ht_for_an`.
+    :param variant_filter_ht: Table with all variants that need AN to be computed.
+    :return: Table with the allele number for new variants in the v4.0 release.
+    """
+    n_samples = group_membership_ht.count()
+    n_groups = len(group_membership_ht.group_membership.take(1)[0])
+
+    vmt = vds.variant_data
+    rmt = vds.reference_data.select_cols().select_rows()
+
+    # Confirm that all variants in variant_filter_ht are present in the
+    # vds.variant_dataset and throw an error if this is not the case since all samples
+    # added to v4.0 genomes should be in this vds.
+    if variant_filter_ht.aggregate(
+        hl.agg.any(hl.is_missing(vmt.rows()[variant_filter_ht.key]))
+    ):
+        raise ValueError(
+            "Not all variants in the variant_filter_ht are found in the "
+            "vds.variant_dataset!"
+        )
+
+    # Filter the VDS variant data and reference data to only keep samples that were in
+    # the v3.1 release. We do it this way instead of using hl.vds.filter_samples because
+    # we want to keep all variants that were in the v3.1 VDS not only those found in the
+    # release and filter_samples includes vmt = vmt.filter_rows(hl.agg.count() > 0).
+    release_s = vmt.aggregate_cols(
+        hl.agg.filter(vmt.meta.release, hl.agg.collect_as_set(vmt.s)), _localize=False
+    )._persist()
+    vmt = vmt.filter_cols(release_s.contains(vmt.s))
+    rmt = rmt.filter_cols(release_s.contains(rmt.s))
+    rmt = rmt.filter_rows(hl.agg.count() > 0)
+
+    # Only keep necessary entries and col annotations for AN calculation from the VDS
+    # so that only the needed entries are densified. AD, DP, GT, and GQ are needed for
+    # adding the adj annotation to compute AN for all adj stratification groups.
+    # sex_karyotype is needed by adjust_sex_ploidy to adjust relevant genotypes on chrX
+    # and chrY non-PAR before computing AN.
+    vmt = (
+        vmt.select_cols(
+            sex_karyotype=vmt.meta.sex_imputation.sex_karyotype,
+            group_membership=group_membership_ht[vmt.col_key].group_membership,
+        )
+        .select_entries("AD", "DP", "GT", "GQ")
+        .select_rows()
+    )
+    vmt = vmt.semi_join_rows(variant_filter_ht)
```

Comment:
doesn't matter where we do it, if you think it is clearer somewhere else, please add as a suggestion. 

`rmt` is only ref blocks not variants so we cant filter to variants in `variant_filter_ht`, and we don't want to filter to only the loci in `variant_filter_ht` because that will get rid of most of the ref block info we need to fill in for computing AN.

Reply:
Oh, I was thinking it might be faster to filter to a smaller set of variant_data, but if it won't speed up, it's fine like this.

Reply:
I could be wrong, but I think Hail's lazy evaluation will handle this for us, so it won't actually perform the entry selection on everything before it filters variants